File: MainWindowTabSubtitleDraft.vb
Imports Subtitle_draft_GMTPC.Models
Imports Subtitle_draft_GMTPC.Services

Partial Class MainWindow

#Region "Subtitle Draft - Fields"

    Private _originalLines As New List(Of SubtitleLine)()
    Private _timeCodeLines As New List(Of SubtitleLine)()
    Private _connectGapLines As New List(Of SubtitleLine)()
    Private _currentFormat As SubtitleFormat = SubtitleFormat.Unknown
    Private _isUpdating As Boolean = False
    Private _totalShiftMs As Integer = 0

#End Region

#Region "Subtitle Draft - Event Handlers"

    Private Sub TxtOriginal_TextChanged(sender As Object, e As TextChangedEventArgs)
        If _isUpdating Then Return
        Try
            _isUpdating = True
            Dim content = SubtitleParser.SanitizeContent(TxtOriginal.Text)
            If String.IsNullOrWhiteSpace(content) Then
                _originalLines.Clear()
                _timeCodeLines.Clear()
                _connectGapLines.Clear()
                TxtTimeCode.Text = ""
                TxtConnectGap.Text = ""
                TxtResult.Text = ""
                TxtOriginalFormat.Text = ""
                _currentFormat = SubtitleFormat.Unknown
                Return
            End If
            _currentFormat = SubtitleParser.DetectFormat(content)
            _originalLines = SubtitleParser.Parse(content)
            TxtOriginalFormat.Text = String.Format("({0} - {1} dòng)", _currentFormat.ToString().ToUpper(), _originalLines.Count)
            _timeCodeLines = New List(Of SubtitleLine)()
            For Each line In _originalLines
                _timeCodeLines.Add(line.Clone())
            Next
            UpdateTimeCodeDisplay()
            _connectGapLines = New List(Of SubtitleLine)()
            For Each line In _timeCodeLines
                _connectGapLines.Add(line.Clone())
            Next
            UpdateConnectGapDisplay()
            UpdateResultDisplay()
        Catch ex As Exception
            TxtOriginalFormat.Text = String.Format("(Lỗi: {0})", ex.Message)
        Finally
            _isUpdating = False
        End Try
    End Sub

    Private Sub BtnTimePlus200_Click(sender As Object, e As RoutedEventArgs)
        If _timeCodeLines.Count = 0 Then Return
        _totalShiftMs += 200
        _timeCodeLines = TimeCodeService.ShiftTime(_timeCodeLines, 200)
        UpdateTimeCodeDisplay()
        _connectGapLines = New List(Of SubtitleLine)()
        For Each line In _timeCodeLines
            _connectGapLines.Add(line.Clone())
        Next
        UpdateConnectGapDisplay()
        UpdateResultDisplay()
        Dim seconds = Math.Abs(_totalShiftMs) / 1000.0
        Dim sign = If(_totalShiftMs >= 0, "+", "-")
        ShowToastDraft(String.Format("⏱️ Đã shift {0}{1}ms ({2}s)", sign, _totalShiftMs, seconds.ToString("0.0")))
    End Sub

    Private Sub BtnTimeMinus200_Click(sender As Object, e As RoutedEventArgs)
        If _timeCodeLines.Count = 0 Then Return
        _totalShiftMs -= 200
        _timeCodeLines = TimeCodeService.ShiftTime(_timeCodeLines, -200)
        UpdateTimeCodeDisplay()
        _connectGapLines = New List(Of SubtitleLine)()
        For Each line In _timeCodeLines
            _connectGapLines.Add(line.Clone())
        Next
        UpdateConnectGapDisplay()
        UpdateResultDisplay()
        Dim seconds = Math.Abs(_totalShiftMs) / 1000.0
        Dim sign = If(_totalShiftMs >= 0, "+", "-")
        ShowToastDraft(String.Format("⏱️ Đã shift {0}{1}ms ({2}s)", sign, _totalShiftMs, seconds.ToString("0.0")))
    End Sub

    Private Sub BtnCopyTimeCode_Click(sender As Object, e As RoutedEventArgs)
        If String.IsNullOrWhiteSpace(TxtTimeCode.Text) Then Return
        Try
            Clipboard.SetText(TxtTimeCode.Text)
            MessageBox.Show("Đã copy!", "Copy", MessageBoxButton.OK, MessageBoxImage.Information)
        Catch ex As Exception
            MessageBox.Show("Lỗi: " & ex.Message, "Lỗi", MessageBoxButton.OK, MessageBoxImage.Error)
        End Try
    End Sub

    Private Sub BtnConnectGap_Click(sender As Object, e As RoutedEventArgs)
        If _connectGapLines.Count = 0 Then Return
        _connectGapLines = TimeCodeService.ConnectGap(_connectGapLines)
        UpdateConnectGapDisplay()
        UpdateResultDisplay()
        ShowToastDraft("🔗 Đã connect gap!")
    End Sub

    Private Sub BtnCopyResult_Click(sender As Object, e As RoutedEventArgs)
        If String.IsNullOrWhiteSpace(TxtResult.Text) Then Return
        Try
            Clipboard.SetText(TxtResult.Text)
            ShowToastDraft("📋 Đã copy Result!")
        Catch ex As Exception
            MessageBox.Show("Lỗi: " & ex.Message, "Lỗi", MessageBoxButton.OK, MessageBoxImage.Error)
        End Try
    End Sub

#End Region

#Region "Subtitle Draft - Display Methods"

    Private Sub UpdateTimeCodeDisplay()
        If _timeCodeLines.Count = 0 Then
            TxtTimeCode.Text = ""
            Return
        End If
        TxtTimeCode.Text = SubtitleParser.ToText(_timeCodeLines, _currentFormat)
    End Sub

    Private Sub UpdateConnectGapDisplay()
        If _connectGapLines.Count = 0 Then
            TxtConnectGap.Text = ""
            Return
        End If
        TxtConnectGap.Text = SubtitleParser.ToText(_connectGapLines, _currentFormat)
    End Sub

    Private Sub UpdateResultDisplay()
        If _connectGapLines.Count > 0 Then
            TxtResult.Text = SubtitleParser.ToText(_connectGapLines, _currentFormat)
        ElseIf _timeCodeLines.Count > 0 Then
            TxtResult.Text = SubtitleParser.ToText(_timeCodeLines, _currentFormat)
        Else
            TxtResult.Text = ""
        End If
    End Sub

#End Region

#Region "Subtitle Draft - Toast"

    Private Async Sub ShowToastDraft(message As String)
        ToastText.Text = message
        ToastBorder.Visibility = Visibility.Visible
        Await Task.Delay(2000)
        ToastBorder.Visibility = Visibility.Collapsed
    End Sub

#End Region

End Class

File: MainWindowTabSubtitleDraft.cs
using System;
using System.Windows;
using System.Windows.Controls;
using System.Threading.Tasks;
using System.Collections.Generic;
using Subtitle_draft_GMTPC.Models;
using Subtitle_draft_GMTPC.Services;

namespace Subtitle_draft_GMTPC
{
    public partial class MainWindow : Window
    {

#region "Subtitle Draft - Fields"

    private List<SubtitleLine> _originalLines = new List<SubtitleLine>();
    private List<SubtitleLine> _timeCodeLines = new List<SubtitleLine>();
    private List<SubtitleLine> _connectGapLines = new List<SubtitleLine>();
    private SubtitleFormat _currentFormat = SubtitleFormat.Unknown;
    private bool _isUpdating = false;
    private int _totalShiftMs = 0;

#endregion

#region "Subtitle Draft - Event Handlers"

    private void TxtOriginal_TextChanged(object sender, TextChangedEventArgs e)
    {
        if (_isUpdating) return;
        try
        {
            _isUpdating = true;
            var content = SubtitleParser.SanitizeContent(TxtOriginal.Text);
            if (string.IsNullOrWhiteSpace(content))
            {
                _originalLines.Clear();
                _timeCodeLines.Clear();
                _connectGapLines.Clear();
                TxtTimeCode.Text = "";
                TxtConnectGap.Text = "";
                TxtResult.Text = "";
                TxtOriginalFormat.Text = "";
                _currentFormat = SubtitleFormat.Unknown;
                return;
            }
            _currentFormat = SubtitleParser.DetectFormat(content);
            _originalLines = SubtitleParser.Parse(content);
            TxtOriginalFormat.Text = string.Format("({0} - {1} dòng)", _currentFormat.ToString().ToUpper(), _originalLines.Count);
            _timeCodeLines = new List<SubtitleLine>();
            foreach (var line in _originalLines)
            {
                _timeCodeLines.Add(line.Clone());
            }
            UpdateTimeCodeDisplay();
            _connectGapLines = new List<SubtitleLine>();
            foreach (var line in _timeCodeLines)
            {
                _connectGapLines.Add(line.Clone());
            }
            UpdateConnectGapDisplay();
            UpdateResultDisplay();
        }
        catch (Exception ex)
        {
            TxtOriginalFormat.Text = string.Format("(Lỗi: {0})", ex.Message);
        }
        finally
        {
            _isUpdating = false;
        }
    }

    private void BtnTimePlus200_Click(object sender, RoutedEventArgs e)
    {
        if (_timeCodeLines.Count == 0) return;
        _totalShiftMs += 200;
        _timeCodeLines = TimeCodeService.ShiftTime(_timeCodeLines, 200);
        UpdateTimeCodeDisplay();
        _connectGapLines = new List<SubtitleLine>();
        foreach (var line in _timeCodeLines)
        {
            _connectGapLines.Add(line.Clone());
        }
        UpdateConnectGapDisplay();
        UpdateResultDisplay();
        var seconds = Math.Abs(_totalShiftMs) / 1000.0;
        var sign = (_totalShiftMs >= 0) ? "+" : "-";
        ShowToastDraft(string.Format("\u23f1\ufe0f Đã shift {0}{1}ms ({2}s)", sign, _totalShiftMs, seconds.ToString("0.0")));
    }

    private void BtnTimeMinus200_Click(object sender, RoutedEventArgs e)
    {
        if (_timeCodeLines.Count == 0) return;
        _totalShiftMs -= 200;
        _timeCodeLines = TimeCodeService.ShiftTime(_timeCodeLines, -200);
        UpdateTimeCodeDisplay();
        _connectGapLines = new List<SubtitleLine>();
        foreach (var line in _timeCodeLines)
        {
            _connectGapLines.Add(line.Clone());
        }
        UpdateConnectGapDisplay();
        UpdateResultDisplay();
        var seconds = Math.Abs(_totalShiftMs) / 1000.0;
        var sign = (_totalShiftMs >= 0) ? "+" : "-";
        ShowToastDraft(string.Format("\u23f1\ufe0f Đã shift {0}{1}ms ({2}s)", sign, _totalShiftMs, seconds.ToString("0.0")));
    }

    private void BtnCopyTimeCode_Click(object sender, RoutedEventArgs e)
    {
        if (string.IsNullOrWhiteSpace(TxtTimeCode.Text)) return;
        try
        {
            Clipboard.SetText(TxtTimeCode.Text);
            System.Windows.MessageBox.Show("Đã copy!", "Copy", System.Windows.MessageBoxButton.OK, System.Windows.MessageBoxImage.Information);
        }
        catch (Exception ex)
        {
            System.Windows.MessageBox.Show("Lỗi: " + ex.Message, "Lỗi", System.Windows.MessageBoxButton.OK, System.Windows.MessageBoxImage.Error);
        }
    }

    private void BtnConnectGap_Click(object sender, RoutedEventArgs e)
    {
        if (_connectGapLines.Count == 0) return;
        _connectGapLines = TimeCodeService.ConnectGap(_connectGapLines);
        UpdateConnectGapDisplay();
        UpdateResultDisplay();
        ShowToastDraft("\ud83d\udd17 Đã connect gap!");
    }

    private void BtnCopyResult_Click(object sender, RoutedEventArgs e)
    {
        if (string.IsNullOrWhiteSpace(TxtResult.Text)) return;
        try
        {
            Clipboard.SetText(TxtResult.Text);
            ShowToastDraft("\ud83d\udccb Đã copy Result!");
        }
        catch (Exception ex)
        {
            System.Windows.MessageBox.Show("Lỗi: " + ex.Message, "Lỗi", System.Windows.MessageBoxButton.OK, System.Windows.MessageBoxImage.Error);
        }
    }

#endregion

#region "Subtitle Draft - Display Methods"

    private void UpdateTimeCodeDisplay()
    {
        if (_timeCodeLines.Count == 0)
        {
            TxtTimeCode.Text = "";
            return;
        }
        TxtTimeCode.Text = SubtitleParser.ToText(_timeCodeLines, _currentFormat);
    }

    private void UpdateConnectGapDisplay()
    {
        if (_connectGapLines.Count == 0)
        {
            TxtConnectGap.Text = "";
            return;
        }
        TxtConnectGap.Text = SubtitleParser.ToText(_connectGapLines, _currentFormat);
    }

    private void UpdateResultDisplay()
    {
        if (_connectGapLines.Count > 0)
        {
            TxtResult.Text = SubtitleParser.ToText(_connectGapLines, _currentFormat);
        }
        else if (_timeCodeLines.Count > 0)
        {
            TxtResult.Text = SubtitleParser.ToText(_timeCodeLines, _currentFormat);
        }
        else
        {
            TxtResult.Text = "";
        }
    }

#endregion

#region "Subtitle Draft - Toast"

    private async void ShowToastDraft(string message)
    {
        ToastText.Text = message;
        ToastBorder.Visibility = Visibility.Visible;
        await Task.Delay(2000);
        ToastBorder.Visibility = Visibility.Collapsed;
    }

#endregion

    }
}
